addons/mail/static/src/core/common/composer.js

```diff
@@ -510,20 +510,18 @@ export class Composer extends Component {
             onClose: (...args) => {
                 // args === [] : click on 'X'
                 // args === { special: true } : click on 'discard'
-                const isDiscard = args.length === 0 || args[0]?.special;
+                const closedWithX = args.length === 0;
+                const isDiscard = closedWithX || args[0]?.special;
                 // otherwise message is posted (args === [undefined])
                 if (!isDiscard && this.props.composer.thread.model === "mail.box") {
                     this.notifySendFromMailbox();
                 }
-                if (
-                    args.length === 0 &&
-                    document
-                        .querySelector(".o_mail_composer_form_view .note-editable")
-                        .innerText.replace(/^\s*$/gm, "")
-                ) {
+                const editor = document.querySelector(".o_mail_composer_form_view .note-editable");
+                const editorIsEmpty = !editor || !editor.innerText.replace(/^\s*$/gm, "");
+                if (closedWithX && !editorIsEmpty) {
                     this.saveContent();
                     this.restoreContent();
-                } else {
+                } else if (!closedWithX || (closedWithX && editor && editorIsEmpty)) {
                     this.clear();
```

Comment:
What about some defensive code in `clear()`?
The core problem of PR is crash in this function, this could be avoided by function being more defensive.
Using heuristics on specific actions for clear puts too much effort in maintaining them, e.g. if some other actions close the composer while not being considered a discard, we'll have to change these LOCs every time.

The condition for `saveContent()` are more likely to be kept exhaustive, that's why it's best to keep the `clear()` as close to a whole fallback as possible.

Reply:
> clear

The issue actually seemed to come from `querySelector(...).innerText` when we couldn't guarantee `querySelector` wouldn't be null.

The changes here are to avoid calling "clear" when escaping from that view, which seems more consistent.

If you prefer the fix can just be adding a couple `?.innerText?.replace`, all it changes is then it might clear when you don't want to.

Reply:
In the general sense, the content is either saved or cleared.
The core aspect of these LOCs is to not loose content, but clear when the message is posted.

These are the most recent specs:
1. ESC is disabled so not an issue
2. Discard button should close the full composer and clear its content
3. X should close the full composer but keep its content (HTML converted to text)
4. Closing tab should recover the content
5. Posting the message clears the content

ESC, X and closing tabs are considered as potentially accidental user actions, so that's why we decide to save content. Discard button is not considered as accidental and the user genuinely wants to drop/clear the message being composed. Clear on posting the message is of course to not have to clear composer manually by hand.

So if I understand the issue correctly, `querySelector(...).innerText` sometimes doesn't work. This is a bit bothersome because this means we cannot save the most up-to-date content from user, so this leads to some loss of content. It'd fine if the previously saved content was recent, the loss is reduced and not a huge deal. This is why we have `saveContentDebounced` that saves the content every 5 seconds while typing text.

-----

- `args.length === 0` is enough to determine there should be no `clear()` (this is a discard).
- These LOCs:
```
document.querySelector(".o_mail_composer_form_view .note-editable")
                 .innerText.replace(/^\s*$/gm, "")
```
Can be put in an inner-branching, so that the save content works only if we can actually retrieve the content for save.
- The `clear()` happens only when posting message or non-accidental discard. This is the opposite of `args.length === 0`. Being able to retrieve some text content in html editor is not a condition for clearance.

------

Hope this gives all the information you need. I think this should be:

```js
if (isDiscard) {
    if (!editorIsEmpty) {
        saveContent();
    }
} else {
    clearContent();
}
```

The `isDiscard` and `editorIsEmpty` leads to neither `saveContent()` nor `clearContent()`. Functionally we'd want to save but we cannot, so we rely on the last saved composer content which hopefully is relatively recent thanks to `saveContentDebounced ` (5 seconds save from most-recent text typed in full composer).

Reply:
@alexkuhn 
Yep, that's the issue and the solution is pretty much what I had in mind.
Refactored as suggested, keeping "args.length === 0" as "accidentalDiscard" to make the block a bit clearer on its own.

The issue should be rare enough, even if potentially very annoying.